matt-1875: provision recipes for flask-gunicorn apps

nmaekawa opened the pull request:
- create non-priviledged user
- install python pkgs and supervisor
- virtualenv and pip requirements
- gunicorn start script virtualenv
- prevent re-create virtualenv
- redis
- logrotate for app related logs
- using refactor for certificates
- (fix pull #81) more descriptive names
- (fix pull #81) string of packages to `install-packages` recipe
- (fix pull #81) redis default save to disk is enough for now
- (fix pull #81) nginx configs
- (fix pull #81) logrotate keeps 7 logs instead of 45

djcp commented on libraries/default.rb:
```diff
@@ -382,6 +382,27 @@ def create_ssl_cert(ssl_info)
       end
     end
+
+    def get_capture_agent_manager_info
```
is this used anywhere?

nmaekawa replied:
i went back and forth but i'm using this now (indirectly) on most capture-agent-manager recipes.

djcp commented on recipes/configure-capture-agent-manager-gunicorn.rb:
```diff
@@ -0,0 +1,30 @@
+app_name = capture_agent_manager_info.fetch(:capture_agent_manager_name, "capture_agent_manager")
+
+execute "install gunicorn" do
+  command "source /home/capture_agent_manager/sites/#{app_name}/venv/bin/activate" \
```
Oh! This works! Neat. I've never thought to do multiline strings like this!

djcp commented on recipes/configure-capture-agent-manager-gunicorn.rb:
```diff
@@ -0,0 +1,30 @@
+  })
+end
+
+directory "/home/capture_agent_manager/sock" do
```
should this be created before the other stuff is run?

nmaekawa replied:
removed this since these dirs are created in a recipe that is run earlier.

djcp commented on recipes/configure-capture-agent-manager-supervisor.rb:
```diff
@@ -0,0 +1,18 @@
+  })
+end
+
+execute %Q|service supervisor restart && supervisorctl start #{app_name}|
```
the restart doesn't automatically start the `app_name` ?  But this still works correctly when you do a `sudo reboot`?

nmaekawa replied:
i think i wasn't sure and run both just in case. i'm sure now since i've tested this a couple of times and it works how is supposed to. removed the chained `supervisor start <app>`.

djcp commented on templates/default/capture-agent-manager-gunicorn-start.sh.erb:
```diff
@@ -0,0 +1,32 @@
+FLASKDIR="/home/capture_agent_manager/sites/${NAME}"
+VENVDIR="${FLASKDIR}/venv"
+SOCKFILE="/home/capture_agent_manager/sock/${NAME}.sock"
+USER=capture_agent_manager
```
should we template this name?

djcp commented:
looks good! There are a few places where you could probably template the name a little more thoroughly.

Where are you thinking you'd document the purpose of this an how to use it? If you follow the pattern @lbjay and I have been following, it'd be in a README in mh-opsworks proper.

nmaekawa commented:
added info on metadata.rb and CHANGELOG.md.
also, lots of little edits to distinguish the capture-agent-manager username from capture-agent-manager app name, and to make it `configurable` -- i think when i started writing the recipes the app would be `redunlive` and i changed to `cadash` to include other related functionalities (and wanted the app name to be configurable because of that)

djcp commented on CHANGELOG.md:
```diff
@@ -2,6 +2,20 @@
+* recipes to install and configure the utility node to host the capture-agent-manager
+  flask-gunicorn webapp
+
+* *REQUIRES EDITS TO THE CLUSTER CONFIG* *REQUIRES MANUAL CHEF RECIPE RUNS*:
```
I would update this language to make sure it's understood this is optional and only for the DCE production cluster.

djcp commented on libraries/default.rb:
```diff
@@ -382,6 +382,40 @@ def create_ssl_cert(ssl_info)
+      ca_info[:capture_agent_manager_app_name]
+    end
+
+    def get_capture_agent_manager_usr_name
```
well, "e" is one of the most common letters in the english language - I guess we're running out? :)

djcp commented on metadata.rb:
```diff
@@ -867,6 +867,101 @@
+this is relevant for the utility node, where the capture-agent-manager `cadash`
+should run.
+
+=== attributes
```
nice! but for consistency it'd be nice to match capitalization of all the other headers (nitpicky, yes).

djcp commented on recipes/configure-capture-agent-manager-gunicorn.rb:
```diff
@@ -0,0 +1,25 @@
+::Chef::Recipe.send(:include, MhOpsworksRecipes::RecipeHelpers)
+
+app_name = get_capture_agent_manager_app_name
+username = get_capture_agent_manager_usr_name
```
"usr_name" and "username". . .

djcp commented:
A couple relatively minor clarifications, otherwise looks good.

lbjay commented on libraries/default.rb:
```diff
@@ -382,6 +382,40 @@ def create_ssl_cert(ssl_info)
+          ldap_base_search: 'dc=some-domain,dc=com',
+          ldap_bind_dn: 'cn=fake_usr,dc=some-domain,dc=com',
+          ldap_bind_passwd: 'pwd',
+          capture_agent_manager_git_repo: 'https://github.com/capture_agent_git_repo',
```
i'm not sure it's necessary to make this configurable. will it change? if you look at the `install-ua-harvester` recipe I just put the repo path right in the code.

nmaekawa replied:
hum.. this is because i was working on a fork.

lbjay commented on recipes/configure-capture-agent-manager-gunicorn.rb:
```diff
@@ -0,0 +1,25 @@
+usr_name = get_capture_agent_manager_usr_name
+
+execute "install gunicorn" do
+  command "source /home/#{usr_name}/sites/#{app_name}/venv/bin/activate" \
```
fyi, not a biggie but you can run `/home/#{usr_name}/sites/#{app_name}/venv/bin/pip` directly and it'll work in the context of that virtualenv without the `source` part

lbjay replied:
Also, is gunicorn not installed via the cadash's requirements.txt?

nmaekawa replied:
yes, it does. how much re-usable a recipe should be? i could write this assuming that the app requirements already run and just install the gunicorn-start script?

lbjay commented on recipes/configure-capture-agent-manager-nginx-proxy.rb:
```diff
@@ -0,0 +1,38 @@
+ssl_info = node.fetch(:ca_ssl, get_dummy_cert)
+if cert_defined(ssl_info)
+  create_ssl_cert(ssl_info)
+  certificate_exists = true
```
It doesn't look like you're using `certificate_exists` anywhere, which makes me wonder if it would ever be the case that ssl would not be enabled? If not you might consider simplifying, e.g., the `install-kibana` recipe, where I just include the dummy cert in the default configuration.

nmaekawa replied:
this app requires ssl because of ldap login... but i think i left this way so when in production i can set a real certificate.

lbjay commented on templates/default/capture-agent-manager-supervisor-conf.erb:
```diff
@@ -0,0 +1,14 @@
+redirect_stderr = True
+
+stdout_logfile = /home/<%= @capture_agent_manager_usr_name %>/logs/<%= @capture_agent_manager_name %>_gunicorn_stdout.log
+stderr_logfile = /home/<%= @capture_agent_manager_usr_name %>/logs/<%= @capture_agent_manager_name %>_gunicorn_stderr.log
```
I don't think you need this since you've got `redirect_stderr = True`. Everything is going to go to `stdout_logfile`.

lbjay commented on templates/default/nginx-proxy-capture-agent-manager.conf.erb:
```diff
@@ -0,0 +1,10 @@
+    location / {
+        proxy_buffering off;
+        proxy_pass http://unix:/home/<%= @capture_agent_manager_usr_name %>/sock/<%= @capture_agent_manager %>.sock;
```
is there a reason you chose sockets vs tcp here?

nmaekawa replied:
i copied from an example i googled... but looks nice that it will not allow anyone to access the app via some tcp port. do you think this has drawbacks?

lbjay commented on templates/default/nginx-proxy-ssl-only.erb:
```diff
@@ -0,0 +1,33 @@
+server {
+    listen 443 default_server ssl;
```
It's pretty simple to listen on :80 + redirect to https too. See `elk-nginx-proxy-conf.erb`

nmaekawa replied:
i was doing that but then dan suggested that we not even redirecting.

lbjay commented on templates/default/capture-agent-manager-gunicorn-start.sh.erb:
```diff
@@ -0,0 +1,32 @@
+export PYTHONPATH=${FLASKIDR}:${PYTHONPATH}
+
+# export all env vars for app
+source ${FLASKDIR}/${NAME}.env
```
I'm not used to seeing `.env` files used in this way, but I think it's probably OK if makes things easier. 

Normally I would expect the `.env` to just list the `VARIABLE=foo` pairs (without the `export`), then your flask app would read/use the settings via something like https://github.com/grauwoelfchen/flask-dotenv

nmaekawa replied:
i think it used to be just variable=foo when it was set up for heroku... what should be extension then? `.sh`? i guess i'm still learning these patterns to setup stuff in the env.

lbjay commented on recipes/configure-capture-agent-manager-supervisor.rb:
```diff
@@ -0,0 +1,19 @@
+
+include_recipe "mh-opsworks-recipes::update-package-repo"
+::Chef::Recipe.send(:include, MhOpsworksRecipes::RecipeHelpers)
+install_package("supervisor")
```
I :heart: supervisord

lbjay commented on recipes/install-capture-agent-manager.rb:
```diff
@@ -0,0 +1,55 @@
+end
+
+execute "install capture_agent_manager dependencies" do
+  command %Q|source /home/#{usr_name}/sites/#{app_name}/venv/bin/activate && pip install -r /home/#{usr_name}/sites/#{app_name}/requirements.txt|
```
Again, you should be able to skip the `source`ing and just exec `/home/#{usr_name}/sites/#{app_name}/venv/bin/pip` directly.

lbjay commented on templates/default/capture-agent-manager-gunicorn-start.sh.erb:
```diff
@@ -0,0 +1,32 @@
+    --name ${NAME} \
+    --workers ${NUM_WORKERS} \
+    --user=${USER} --group=${GROUP} \
+    --log-level=debug \
```
should this be a config setting in `get_capture_agent_manager_info`?

lbjay commented:
:+1:
